### PR TITLE
EOS-14151:m0kv_metadata_parsing_tool

### PR DESCRIPTION
**Pull request by SaumitraK531**
Signed-off-by: SaumitraK531 saumitra.kulkarni@seagate.com

### CODE REVIEW

File: m0kv_metadata_parsing_tool/m0kv_metadata_parsing_tool.sh [hunk 1]
```diff
@@ -0,0 +1,417 @@
+red=`tput setaf 1`
+green=`tput setaf 2`
+reset=`tput sgr0`
+rm base64_encoder_decoder -f /dev/null 2>&1
+g++ base64_encoder_decoder.cc -o base64_encoder_decoder -std=c++11 > /dev/null 2>&1
```
Comment (nileshgovande) [inline m0kv_metadata_parsing_tool/m0kv_metadata_parsing_tool.sh:5]: Write a small makefile

**Review by nileshgovande — CHANGES_REQUESTED** [review]

File: m0kv_metadata_parsing_tool/m0kv_metadata_parsing_tool.sh [hunk 2]
```diff
@@ -0,0 +1,417 @@
...
+        echo -e "\n${green}##########################################################################################${reset}"
+        rm /tmp/m0kvtest1.log > /dev/null 2>&1
+        rm *.[0-9]* -f > /dev/null 2>&1 
+    else
+        binary_error
+        break
+    fi
+}
+
+while getopts "l:h:p:f:n:b:d:" opt
```
Comment (nileshgovande) [inline m0kv_metadata_parsing_tool/m0kv_metadata_parsing_tool.sh:373]: Make it more browsable and maintainable

File: m0kv_metadata_parsing_tool/README.md [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+1. Requires m0kv binary. 
+* run build_motr.sh in 'third_party/build_motr.sh'
+* m0kv binary is generated here 'third_party/motr/utils/m0kv'
+
+
+
+## Usage
+
+```
+./m0kv_metadata_parsing_tool.sh -l 10.230.247.176@tcp:12345:33:905 -h 10.230.247.176@tcp:12345:34:1 -p '<0x7000000000000001:0>' -f '<0x7200000000000000:0>' -b 0 -d 10
```
Comment (nileshgovande) [inline m0kv_metadata_parsing_tool/README.md:18]: m0kv_parsing_tool -l '$local_addr' -h '$ha_addr' -p '$profile' -f '$proc_fid' -s index

File: m0kv_metadata_parsing_tool/README.md [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+-----------------------------------------------------------------------------------
+-b is flag and required arguement (can be set 0 or 1)
```
Comment (nileshgovande) [inline m0kv_metadata_parsing_tool/README.md:20]: call it d - (depth)

Comment (nileshgovande) [reply m0kv_metadata_parsing_tool/README.md:20]: depth = 1 and 2 (>1)

File: m0kv_metadata_parsing_tool/README.md (hunk 3 above)
Comment (nileshgovande) [reply m0kv_metadata_parsing_tool/README.md:18]: then pls give examples

File: m0kv_metadata_parsing_tool/m0kv_metadata_parsing_tool.sh [hunk 5]
```diff
@@ -0,0 +1,417 @@
+red=`tput setaf 1`
+green=`tput setaf 2`
+reset=`tput sgr0`
+rm base64_encoder_decoder -f /dev/null 2>&1
+g++ base64_encoder_decoder.cc -o base64_encoder_decoder -std=c++11 > /dev/null 2>&1
+
+m0kv_PATH="third_party/motr/utils/m0kv"
+
+usage() {
+    echo "${red}Usage:${green} $0 -l 10.230.247.176@tcp:12345:33:905 -h 10.230.247.176@tcp:12345:34:1 -p '<0x7000000000000001:0>' -f '<0x7200000000000000:0>' -b 0 -d 10${reset}"
```
Comment (nileshgovande) [inline m0kv_metadata_parsing_tool/m0kv_metadata_parsing_tool.sh:10]: same comment as readme

File: m0kv_metadata_parsing_tool/m0kv_metadata_parsing_tool.sh [hunk 6]
```diff
@@ -0,0 +1,417 @@
...
+    re='^[0-9]+$'
+    if [ $BUCKET_LEVEL_FLAG -le 1 ] && [[ $BUCKET_LEVEL_FLAG =~ $re ]] && [[ $NO_OF_DEL =~ $re ]]
+    then
+        get_metadata_of_bucket $BUCKET_NAME $L $H $P $F $BUCKET_LEVEL_FLAG
+        get_probable_delete_index $L $H $P $F $NO_OF_DEL
+        rm base64_encoder_decoder -f /dev/null 2>&1
+        echo -e "\n${red}Log file :${reset} ${green}/var/log/m0kv_metadata.log${reset}\n"
+    else
+        usage
+        rm base64_encoder_decoder -f /dev/null 2>&1
```
Comment (nileshgovande) [inline m0kv_metadata_parsing_tool/m0kv_metadata_parsing_tool.sh:411]: Put only once - in the end

File: m0kv_metadata_parsing_tool/Makefile [hunk 7]
```diff
@@ -0,0 +1,5 @@
+COMPILER=g++
+all:
+	$(COMPILER) base64_encoder_decoder.cc -o base64_encoder_decoder -std=c++11
```
Comment (amitwac2608) [inline m0kv_metadata_parsing_tool/Makefile:3]: use -Wall flag to catch any compiler warnings.

File: m0kv_metadata_parsing_tool/base64_encoder_decoder.cc [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+}
+
+
+
+/**********************************************
+this function decodes the base64 encoded string 
+**********************************************/
+std::string base64_decode(std::string const& encoded_string) {
+  int in_len = encoded_string.size();
+  int i = 0;
```
Comment (amitwac2608) [inline m0kv_metadata_parsing_tool/base64_encoder_decoder.cc:51]: do not use 'i', 'j' for iteration. Use proper names for easy code reading.

File: m0kv_metadata_parsing_tool/base64_encoder_decoder.cc [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+  int j = 0;
+  int in_ = 0;
+  unsigned char char_array_4[4], char_array_3[3];
+  std::string ret;
+
+  while (in_len-- && ( encoded_string[in_] != '=') && is_base64(encoded_string[in_])) {
```
Comment (amitwac2608) [inline m0kv_metadata_parsing_tool/base64_encoder_decoder.cc:57]: write what is being done in this logic as comment.

File: m0kv_metadata_parsing_tool/base64_encoder_decoder.cc [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+  return base64_encode((unsigned char const *)&u_hi, sizeof(u_hi)) + "-" +
+         base64_encode((unsigned char const *)&u_lo, sizeof(u_lo));
+}
+
+/****************************************
+main function 
+****************************************/
+int main(int argc, char *argv[])
+{
+  if (std::string(argv[1]) == "-d") {
```
Comment (amitwac2608) [inline m0kv_metadata_parsing_tool/base64_encoder_decoder.cc:175]: What if I do not pass any argument ? In that case your argv array would be of size 1, so I guess accessing argv[1] would give core dump !!!

Check for size of argv array, if its size is greater than 1 or not.

File: m0kv_metadata_parsing_tool/m0kv_metadata_parsing_tool.sh [hunk 11]
```diff
@@ -0,0 +1,428 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# For any questions about this software or licensing,
+# please email opensource@seagate.com or cortx-questions@seagate.com.
+#
+
+red=`tput setaf 1`
```
Comment (amitwac2608) [inline m0kv_metadata_parsing_tool/m0kv_metadata_parsing_tool.sh:20]: use $(cmd), as using back-ticks is not recommended nowdays.
red=$(tput setaf 1)

File: m0kv_metadata_parsing_tool/m0kv_metadata_parsing_tool.sh [hunk 12]
```diff
@@ -0,0 +1,428 @@
...
+green=`tput setaf 2`
+reset=`tput sgr0`
+rm base64_encoder_decoder -f /dev/null 2>&1
+make > /dev/null 2>&1
+m0kv_PATH="third_party/motr/utils/m0kv"
```
Comment (amitwac2608) [inline m0kv_metadata_parsing_tool/m0kv_metadata_parsing_tool.sh:25]: is this tool for dev -environments ?

On CORTX stack, there would not be any path like third_party/..

File: m0kv_metadata_parsing_tool/m0kv_metadata_parsing_tool.sh [hunk 13]
```diff
@@ -0,0 +1,428 @@
...
+    if [[ ! -z "$m0kv_flag" ]]
+    then
+        declare -a KEYS
+        declare -a OBJECTS
+        declare -a OBJECT_LIST_INDEX_OID
+        declare -a OBJECT_VERSION_LIST_INDEX_OID
+        declare -a PART_LIST_INDEX_OID
+        rm /tmp/m0kvprobable.log > /dev/null 2>&1
+        cd ..
+        ./$m0kv_PATH -l $1 -h $2 -p $3 -f $4 index next '0x7800000000000000:0x100003' '0' 100 -s >> /tmp/m0kvprobable.log
```
Comment (amitwac2608) [inline m0kv_metadata_parsing_tool/m0kv_metadata_parsing_tool.sh:62]: declare all the index ids as constant

**Review by amitwac2608 — COMMENTED** [review]
gave some comments.

Also can you attach a sample output of your script to this PR for reference ?

File: m0kv_metadata_parsing_tool/Makefile (hunk 7 above)
Comment (pranavPawar07) [reply m0kv_metadata_parsing_tool/Makefile:3]: Hi Amit, sure will do that.

**Review by nileshgovande — APPROVED** [review]

**Review by amitwac2608 — APPROVED** [review]
looks good

**Comment by nileshgovande** [comment]
No Jenkins needed as it is a separate utility - 2 approvals - merging.

**Pull request merged**